File: BlinkCard/ios/Serialization/BlinkCardDeserializationUtils.swift
//
//  BlinkCardDeserializationUtils.swift
//  BlinkCardReactNative
//
//  Created by Milan Parađina on 10.03.2026..
//

import Foundation
import UIKit
import BlinkCard
import BlinkCardUX

struct BlinkCardDeserializationUtils {
    
    static func deserializeSdkSettings(_ sdkSettingsDict: [String: Any]) throws -> BlinkCardSdkSettings {
        guard let licenseKey = sdkSettingsDict["licenseKey"] as? String, !licenseKey.isEmpty else {
            throw NSError(domain: "", code: 01)
        }
        
        var blinkCardSdkSettings = BlinkCardSdkSettings(licenseKey: licenseKey)
        
        if let licensee = sdkSettingsDict["licensee"] as? String {
            blinkCardSdkSettings.licensee = licensee
        }
        
        if let helloLog = sdkSettingsDict["helloLogEnabled"] as? Bool {
            blinkCardSdkSettings.helloLogEnabled = helloLog
        }
        
        if let downloadResources = sdkSettingsDict["downloadResources"] as? Bool {
            blinkCardSdkSettings.downloadResources = downloadResources
        }
        
        if let resourceDownloadUrl = sdkSettingsDict["resourceDownloadUrl"] as? String {
            blinkCardSdkSettings.resourceDownloadUrl = resourceDownloadUrl
        }
        
        if let resourceLocalFolder = sdkSettingsDict["resourceLocalFolder"] as? String {
            blinkCardSdkSettings.resourceLocalFolder = resourceLocalFolder
        }
        
        if let bundleIdentifier = sdkSettingsDict["bundleIdentifier"] as? String,
            let bundle: Bundle = Bundle.init(identifier: bundleIdentifier) {
            blinkCardSdkSettings.bundleURL = bundle.bundleURL
        }
        
        if let microblinkProxyUrl = sdkSettingsDict["microblinkProxyURL"] as? String {
            blinkCardSdkSettings.microblinkProxyURL = microblinkProxyUrl
        }
        
        return blinkCardSdkSettings
    }
    
    static func deserializeSessionSettings(_ sessionSettingsDict: [String: Any]) -> BlinkCardSessionSettings {
        var sessionSettings = BlinkCardSessionSettings()
        
        if let stepTimeoutDuration = sessionSettingsDict["stepTimeTimeoutInterval"] as? Int {
            sessionSettings.stepTimeoutDuration = TimeInterval(stepTimeoutDuration / 1000)
        }
        
        if let scanningSettings = sessionSettingsDict["scanningSettings"] as? [String: Any] {
            sessionSettings.scanningSettings = deserializeScanningSettings(scanningSettingsDict: scanningSettings)
        }
        
        return sessionSettings
    }
    
    static func deserializeScanningSettings(scanningSettingsDict: [String: Any]) -> ScanningSettings {
        
        var scanningSettings = ScanningSettings()
        
        if let skipImagesWithBlur = scanningSettingsDict["skipImagesWithBlur"] as? Bool {
            scanningSettings.skipImagesWithBlur = skipImagesWithBlur
        }
        
        if let tiltDetectionLevel = scanningSettingsDict["tiltDetectionLevel"] as? String {
            
            scanningSettings.tiltDetectionLevel = deserializeDetectionLevel(tiltDetectionLevel)
        }
        
        if let inputImageMargin = scanningSettingsDict["inputImageMargin"] as? Double {
            scanningSettings.inputImageMargin = Float(inputImageMargin)
        }
        
        if let livenessSettings = scanningSettingsDict["livenessSettings"] as? [String: Any] {
            scanningSettings.livenessSettings = deserializeLivenessSettings(livenessSettingsDict: livenessSettings)
        }
        
        if let extractionSettings = scanningSettingsDict["extractionSettings"] as? [String: Any] {
            scanningSettings.extractionSettings = deserializeExtractionSettings(extractionSettings)
        }
        
        if let croppedImageSettings = scanningSettingsDict["croppedImageSettings"] as? [String: Any] {
            scanningSettings.croppedImageSettings = deserializeCroppedImageSettings(croppedImageSettings)
        }
        
        if let anonymizationSettings = scanningSettingsDict["anonymizationSettings"] as? [String: Any] {
            scanningSettings.anonymizationSettings = deserializeAnonymizationSettings(anonymizationSettings)
        }
        
        return scanningSettings
    }
    
    static func deserializeLivenessSettings(livenessSettingsDict: [String: Any]) -> LivenessSettings {
                
        if let handToCardSizeRatio = livenessSettingsDict["handToCardSizeRatio"] as? Double,
           let handCardOverlapThreshold = livenessSettingsDict["handCardOverlapThreshold"] as? Double,
           let enableCardHeldInHandCheck = livenessSettingsDict["enableCardHeldInHandCheck"] as? Bool,
           let screenCheckStrictnessLevel = livenessSettingsDict["screenCheckStrictnessLevel"] as? String,
           let photocopyCheckStrictnessLevel = livenessSettingsDict["photocopyCheckStrictnessLevel"] as? String {
            return LivenessSettings(
                handToCardSizeRatio: Float(handToCardSizeRatio),
                handCardOverlapThreshold: Float(handCardOverlapThreshold),
                enableCardHeldInHandCheck: enableCardHeldInHandCheck,
                screenCheckStrictnessLevel: deserializeStrictnessLevel(screenCheckStrictnessLevel),
                photocopyCheckStrictnessLevel: deserializeStrictnessLevel(photocopyCheckStrictnessLevel))
        }
        
        return LivenessSettings()
    }
    
    static func deserializeCroppedImageSettings(_ croppedImageSettingsDict: [String: Any]) -> CroppedImageSettings {
        var croppedImageSettings = CroppedImageSettings()
        
        if let dotsPerInch = croppedImageSettingsDict["dotsPerInch"] as? Int {
            croppedImageSettings.dotsPerInch = UInt16(dotsPerInch)
        }
        
        if let extensionFactor = croppedImageSettingsDict["extensionFactor"] as? Double {
            croppedImageSettings.extensionFactor = Float(extensionFactor)
        }
        
        if let returnCardImage = croppedImageSettingsDict["returnCardImage"] as? Bool {
            croppedImageSettings.returnCardImage = returnCardImage
        }
        
        return croppedImageSettings
    }
    
    static func deserializeExtractionSettings(_ extractionSettingsDict: [String: Any]) -> ExtractionSettings {
        
        if let extractIban = extractionSettingsDict["extractIban"] as? Bool,
           let extractExpiryDate = extractionSettingsDict["extractExpiryDate"] as? Bool,
           let extractCardholderName = extractionSettingsDict["extractCardholderName"] as? Bool,
           let extractCvv = extractionSettingsDict["extractCvv"] as? Bool,
           let extractInvalidCardNumber = extractionSettingsDict["extractInvalidCardNumber"] as? Bool {
            return ExtractionSettings(
                extractIban: extractIban,
                extractExpiryDate: extractExpiryDate,
                extractCardholderName: extractCardholderName,
                extractCvv: extractCvv,
                extractInvalidCardNumber: extractInvalidCardNumber)
        }

        return ExtractionSettings()
    }
    
    static func deserializeAnonymizationSettings(_ anonymizationSettingsDict: [String: Any]) -> AnonymizationSettings {
        var anonymizationSettings = AnonymizationSettings()
        
        if let cardNumberAnonymizationSettings = anonymizationSettingsDict["cardNumberAnonymizationSettings"] as? [String: Any] {
            
            anonymizationSettings.cardNumberAnonymizationSettings = deserializeCardNumberAnonymizationSettings(cardNumberAnonymizationSettings)
        }
        
        if let cardNumberPrefixAnonymizationMode = anonymizationSettingsDict["cardNumberPrefixAnonymizationMode"] as? String,
           let mode = AnonymizationMode(rawValue: cardNumberPrefixAnonymizationMode) {
            anonymizationSettings.cardNumberPrefixAnonymizationMode = mode
        }
        
        if let cvvAnonymizationMode = anonymizationSettingsDict["cvvAnonymizationMode"] as? String,
           let mode = AnonymizationMode(rawValue: cvvAnonymizationMode) {
            anonymizationSettings.cvvAnonymizationMode = mode
        }
        
        if let ibanAnonymizationMode = anonymizationSettingsDict["ibanAnonymizationMode"] as? String,
           let mode = AnonymizationMode(rawValue: ibanAnonymizationMode) {
            anonymizationSettings.ibanAnonymizationMode = mode
        }
        
        if let cardholderNameAnonymizationMode = anonymizationSettingsDict["cardholderNameAnonymizationMode"] as? String,
           let mode = AnonymizationMode(rawValue: cardholderNameAnonymizationMode) {
            anonymizationSettings.cardholderNameAnonymizationMode = mode
        }
        
        return anonymizationSettings
    }
    
    static func deserializeCardNumberAnonymizationSettings(_ cardNumberAnonymizationSettingsDict: [String: Any]) -> CardNumberAnonymizationSettings {
        var cardNumberAnonymizationSettings = CardNumberAnonymizationSettings()
        
        if let anonymizationMode = cardNumberAnonymizationSettingsDict["anonymizationMode"] as? String,
        let mode = AnonymizationMode(rawValue: anonymizationMode) {
            cardNumberAnonymizationSettings.mode = mode
        }
        
        if let prefixDigitsVisible = cardNumberAnonymizationSettingsDict["prefixDigitsVisible"] as? Int {
            cardNumberAnonymizationSettings.prefixDigitsVisible = UInt8(prefixDigitsVisible)
        }
        
        if let suffixDigitsVisible = cardNumberAnonymizationSettingsDict["suffixDigitsVisible"] as? Int {
            cardNumberAnonymizationSettings.suffixDigitsVisible = UInt8(suffixDigitsVisible)
        }

        return cardNumberAnonymizationSettings
    }
    
    static func deserializeScanningUxSettings(_ scanningUxSettingsDict: [String: Any]) -> ScanningUXSettings {
        
        if let showIntroductionAlert = scanningUxSettingsDict["showIntroductionAlert"] as? Bool,
           let showHelpButton = scanningUxSettingsDict["showHelpButton"] as? Bool,
           let cameraPosition = scanningUxSettingsDict["preferredCameraPosition"] as? String,
        let allowHapticFeedback = scanningUxSettingsDict["allowHapticFeedback"] as? Bool {
            return ScanningUXSettings(
                showIntroductionAlert: showIntroductionAlert,
                showHelpButton: showHelpButton,
                preferredCameraPosition: deserializeCameraPosition(cameraPosition),
                allowHapticFeedback: allowHapticFeedback)
        }
        
        return ScanningUXSettings()
        
    }
    
    static func deserializeDetectionLevel(_ levelString: String) -> DetectionLevel {
        switch levelString {
        case "off": return .off
        case "low": return .low
        case "mid": return .mid
        case "high": return .high
        default: return .mid
        }
    }
    
    static func deserializeStrictnessLevel(_ levelString: String) -> StrictnessLevel {
        switch levelString {
        case "disabled": return .disabled
        case "level1": return .level1
        case "level2": return .level2
        case "level3": return .level3
        case "level4": return .level4
        case "level5": return .level5
        case "level6": return .level6
        case "level7": return .level7
        case "level8": return .level8
        case "level9": return .level9
        case "level10": return .level10
        default: return .level5
        }
    }
    
    static func deserializeCameraPosition(_ positingString: String) -> Camera.CameraPosition {
        switch positingString {
        case "front": return .front
        case "back": return .back
        default: return .back
        }
    }
    
    static func deserializeBase64Image(_ base64Image: String?) -> UIImage? {
        if let base64Image,
           let imageData = Data(base64Encoded: base64Image, options: .ignoreUnknownCharacters) {
            return UIImage(data: imageData)
        }
        return nil
    }
}

File: BlinkCard/android/src/main/java/com/microblink/blinkcard/reactnative/serialization/BlinkCardDeserializationUtils.kt
package com.microblink.blinkcard.reactnative.serialization

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import com.microblink.blinkcard.core.BlinkCardSdkSettings
import com.microblink.blinkcard.core.session.BlinkCardSessionSettings
import com.microblink.blinkcard.core.session.InputImageSource
import com.microblink.blinkcard.core.settings.AnonymizationMode
import com.microblink.blinkcard.core.settings.AnonymizationSettings
import com.microblink.blinkcard.core.settings.CardNumberAnonymizationSettings
import com.microblink.blinkcard.core.settings.CroppedImageSettings
import com.microblink.blinkcard.core.settings.DetectionLevel
import com.microblink.blinkcard.core.settings.ExtractionSettings
import com.microblink.blinkcard.core.settings.LivenessSettings
import com.microblink.blinkcard.core.settings.ScanningSettings
import com.microblink.blinkcard.core.settings.StrictnessLevel
import com.microblink.blinkcard.core.utils.defaultResourceDownloadUrl
import com.microblink.blinkcard.core.utils.defaultResourcesLocalFolder
import com.microblink.blinkcard.reactnative.BlinkCardReactNativeError
import com.microblink.blinkcard.ux.camera.CameraLensFacing
import com.microblink.blinkcard.ux.camera.CameraSettings
import com.microblink.blinkcard.ux.settings.BlinkCardUxSettings
import org.json.JSONArray
import org.json.JSONObject
import kotlin.time.Duration

object BlinkCardDeserializationUtils {
  fun deserializeBlinkCardSdkSettings(blinkCardSdkSettingsJson: String?): BlinkCardSdkSettings {
      if (blinkCardSdkSettingsJson.isNullOrEmpty()) throw BlinkCardReactNativeError.InvalidSettingsProvided("SDK settings")
    val blinkCardSdkSettingsJson = JSONObject(blinkCardSdkSettingsJson).toMap()

    val licenseKey = blinkCardSdkSettingsJson["licenseKey"] as? String ?: throw BlinkCardReactNativeError.InvalidLicenseKeyProvided()
    return BlinkCardSdkSettings(
      licenseKey = licenseKey,
      licensee = blinkCardSdkSettingsJson["licensee"] as? String,
      downloadResources = blinkCardSdkSettingsJson["downloadResources"] as? Boolean ?: true,
      resourceDownloadUrl = blinkCardSdkSettingsJson["resourceDownloadUrl"] as? String ?: defaultResourceDownloadUrl,
      resourceLocalFolder = blinkCardSdkSettingsJson["resourceLocalFolder"] as? String ?: defaultResourcesLocalFolder,
      microblinkProxyUrl = blinkCardSdkSettingsJson["microblinkProxyURL"] as? String
    )
  }

  fun deserializeBlinkCardSessionSettings(blinkCardSessionSettingsJson: String?, isDirectApi: Boolean = false): BlinkCardSessionSettings {
    val json = blinkCardSessionSettingsJson ?: return BlinkCardSessionSettings()
    val blinkCardSessionSettingsJson = JSONObject(json).toMap()

    return BlinkCardSessionSettings(
      inputImageSource = if (isDirectApi) InputImageSource.Photo else InputImageSource.Video,
      scanningSettings =  deserializeScanningSettings(blinkCardSessionSettingsJson["scanningSettings"] as? Map<*, *>),

      )
  }
  fun deserializeScanningSettings(scanningSettingsJson: Map<*, *>?): ScanningSettings {
    val scanningSettingsJson = scanningSettingsJson ?: return ScanningSettings()
    return ScanningSettings(
      skipImagesWithBlur = scanningSettingsJson["skipImagesWithBlur"] as? Boolean ?: true,
      tiltDetectionLevel = deserializeDetectionLevel(scanningSettingsJson["skipImagesWithBlur"] as? String ?: "mid"),
      inputImageMargin = scanningSettingsJson["skipImagesWithBlur"] as? Float ?: 0.02f,
      livenessSettings = deserializeLivenessSettings(scanningSettingsJson["livenessSettings"] as? Map<*, *>),
      extractionSettings = deserializeExtractionSettings(scanningSettingsJson["extractionSettings"] as? Map<*, *>),
      croppedImageSettings = deserializeCroppedImageSettings(scanningSettingsJson["croppedImageSettings"] as? Map<*, *>),
      anonymizationSettings = deserializeAnonymizationSettings(scanningSettingsJson["anonymizationSettings"] as? Map<*, *>),
    )
  }

  fun deserializeLivenessSettings(livenessSettingsJson: Map<*, *>?): LivenessSettings {
    val livenessSettingsJson = livenessSettingsJson ?: return LivenessSettings()
    return LivenessSettings(
      enableCardHeldInHandCheck = livenessSettingsJson["enableCardHeldInHandCheck"] as? Boolean ?: true,
      handCardOverlapThreshold = livenessSettingsJson["handCardOverlapThreshold"] as? Float ?: 0.05f,
      handToCardSizeRatio = livenessSettingsJson["handToCardSizeRatio"] as? Float ?: 0.15f,
      photocopyCheckStrictnessLevel = deserializeStrictnessLevel(livenessSettingsJson["photocopyCheckStrictnessLevel"] as? String ?: "level5"),
      screenCheckStrictnessLevel = deserializeStrictnessLevel(livenessSettingsJson["screenCheckStrictnessLevel"] as? String ?: "level5")
    )
  }

  fun deserializeExtractionSettings(extractionSettingsJson: Map<*,*>?): ExtractionSettings {
    val extractionSettingsJson = extractionSettingsJson ?: return ExtractionSettings()
    return ExtractionSettings(
      extractIban = extractionSettingsJson["extractIban"] as? Boolean ?: true,
      extractExpiryDate = extractionSettingsJson["extractExpiryDate"] as? Boolean ?: true,
      extractCardholderName = extractionSettingsJson["extractCardholderName"] as? Boolean
        ?: true,
      extractCvv = extractionSettingsJson["extractCvv"] as? Boolean ?: true,
      extractInvalidCardNumber = extractionSettingsJson["extractInvalidCardNumber"] as? Boolean
        ?: false
    )
  }

  fun deserializeCroppedImageSettings(croppedImageSettingsJson: Map<*, *>?): CroppedImageSettings {
    val croppedImageSettingsJson = croppedImageSettingsJson ?: return CroppedImageSettings()
    return CroppedImageSettings(
      dotsPerInch = croppedImageSettingsJson["dotsPerInch"] as? UShort ?: 250u,
      extensionFactor = croppedImageSettingsJson["extensionFactor"] as? Float ?: 0.0f,
      returnCardImage = croppedImageSettingsJson["returnCardImage"] as? Boolean ?: false,
    )
  }

  fun deserializeAnonymizationSettings(anonymizationSettingsJson: Map<*, *>?): AnonymizationSettings {
    val anonymizationSettingsJson = anonymizationSettingsJson ?: return AnonymizationSettings()

    return AnonymizationSettings(
      cardholderNameAnonymizationMode = deserializeAnonymizationMode(anonymizationSettingsJson["cardholderNameAnonymizationMode"] as? String ?: "none"),
      cardNumberPrefixAnonymizationMode = deserializeAnonymizationMode(anonymizationSettingsJson["cardNumberPrefixAnonymizationMode"] as? String ?: "none"),
      cvvAnonymizationMode = deserializeAnonymizationMode(anonymizationSettingsJson["cvvAnonymizationMode"] as? String ?: "none"),
      ibanAnonymizationMode = deserializeAnonymizationMode(anonymizationSettingsJson["ibanAnonymizationMode"] as? String ?: "none"),
      cardNumberAnonymizationSettings = deserializeCardNumberAnonymizationSettings(anonymizationSettingsJson["cardNumberAnonymizationSettings"] as? Map<*, *>)
    )
  }

  fun deserializeCardNumberAnonymizationSettings(cardNumberAnonymizationSettingsJson: Map<*, *>?): CardNumberAnonymizationSettings {
    val cardNumberAnonymizationSettingsJson = cardNumberAnonymizationSettingsJson ?: return CardNumberAnonymizationSettings()
    return CardNumberAnonymizationSettings(
      anonymizationMode = deserializeAnonymizationMode(cardNumberAnonymizationSettingsJson["anonymizationMode"] as? String ?: "none"),
      prefixDigitsVisible = (cardNumberAnonymizationSettingsJson["prefixDigitsVisible"] as? Int ?: 0).toUByte(),
      suffixDigitsVisible = (cardNumberAnonymizationSettingsJson["suffixDigitsVisible"] as? Int ?: 0).toUByte(),
    )
  }

  fun deserializeScanningUxSettings(stepTimeoutDuration: Duration, allowHapticFeedback: Boolean): BlinkCardUxSettings {
    return BlinkCardUxSettings(
      stepTimeoutDuration = stepTimeoutDuration,
      allowHapticFeedback = allowHapticFeedback
    )
  }

  fun deserializeDetectionLevel(level: String): DetectionLevel = DetectionLevel.valueOf(level.replaceFirstChar { it.uppercase() })
  fun deserializeStrictnessLevel(level: String): StrictnessLevel = StrictnessLevel.valueOf(level.replaceFirstChar { it.uppercase() })
  fun deserializeAnonymizationMode(mode: String): AnonymizationMode = AnonymizationMode.valueOf(mode.replaceFirstChar { it.uppercase() })

  fun deserializeCameraSettings(lens: String): CameraSettings  {
    val cameraLens = if (lens == "front") CameraLensFacing.LensFacingFront else CameraLensFacing.LensFacingBack
    return CameraSettings(lensFacing = cameraLens)
  }

  fun base64ToBitmap(base64Str: String?): Bitmap? {
    return try {
      val decodedBytes = Base64.decode(base64Str, Base64.DEFAULT)
      BitmapFactory.decodeByteArray(decodedBytes, 0, decodedBytes.size)
    } catch (_: IllegalArgumentException) {
      null
    }
  }
}

fun JSONObject.toMap(): Map<String, Any?> {
  val map = mutableMapOf<String, Any?>()

  keys().forEach { key ->
    val value = get(key)
    map[key] = when (value) {
      is JSONArray -> value.toList()
      is JSONObject -> value.toMap()
      JSONObject.NULL -> null
      else -> value
    }
  }

  return map
}

fun JSONArray.toList(): List<Any?> {
  val list = mutableListOf<Any?>()

  for (i in 0 until length()) {
    val value = get(i)
    list.add(
      when (value) {
        is JSONArray -> value.toList()
        is JSONObject -> value.toMap()
        JSONObject.NULL -> null
        else -> value
      }
    )
  }

  return list
}
